### ISSUE TITLE
ADDomainController: Add the ability to transfer/seize the FSMO roles

### ISSUE THREAD

**Issue by mprahl** [PowerShell/ActiveDirectoryDsc#55]
Hello,
Being able to manage the FSMO roles via transfer or seizure would be beneficial in the xADDomainController resource. That way if you need to rebuild a Domain Controller, it can be configured entirely via DSC without manual steps at the end of the configuration being applied.

**Comment by iainbrighton** [comment]
@PrahlM93 I can see a reason for being able to specify with node should hold a FSMO role. I can't really see a situation that you'd want to seize a role though. Could you elaborate on when/why you would use this?

**Comment by PrahlM93** [comment]
@iainbrighton, I was thinking if a Domain Controller is destroyed or becomes corrupt, you'd be able to rebuild it from the MOF file. Since the Domain Controller which contained the FSMO roles is now offline because it was destroyed or corrupt, you would have to seize the FSMO roles instead of being able to transfer them.

**Comment by iainbrighton** [comment]
@PrahlM93 I think a resource to specify the FSMO role placement might be in order, seizing the role if necessary, might be better option. My only concern would be that you could easily specify the same FSMO role on multiple nodes and things could get messy very quickly! @GoateePFE @HemantMahawar what are your thoughts?

**Comment by GoateePFE** [comment]
Hmm. Yes, FSMO shifting in a loop would not be cool. I see the value in the resource. This would take some planning. Maybe some kind of check on the FSMO role holder attributes to see if they have been updated lately as a dampening technique to keep them from swapping.  Maybe you could enumerate the current FSMOs, check their DSC configurations to make sure they do not have the same FSMO role set in DSC, and then make the change. We can brainstorm through this.

**Comment by aig79** [comment]
I am new to DSC, and do not know much about Active Directory internals. I do know the FSMO Roles are important and it does not matter what server holds them, also there is clean up work to do when a domain control has to be rebuilt, and was not demoted; remove DNS records, seizing roles, I have been thinking about building configurations for my current environment. Now I am also thinking about migrating to Windows Server 2016. 

User Story 1:
Creating new domain controllers with FSMO roles assigned to the new servers once. 
Then decommission Windows Server 2012R2 Domain Controls.

I have one domain. So having to manually complete the change over is not critical to me, which I will have to lookup all the steps, for each upgrade.

User Story 2:
On seizing the FSMO roles, I would think that would be a new configuration, If the previous server is really dead, or we really need new ids for new records and that role is offline and we shot that server then get the roles and configure a new server to replace it. 

I thought I would add my user story that might guide some ideas. I would also be interested in one configuration not to reapply, but the other configurations in the document to be reapplied.

**Comment by johlju** [comment]
I thought of this a couple of days, and not sure a new resource is the right way to go. I can only see a reason to move Flexible Single Master Operation (FSMO) roles is to spread them to another domain controller when that domain controller is provisioned. 
To move a FSMO role the DC that holds the role must be up and responding, that rules out moving roles to another domain controller to seamless take over the responsibility if the prior holder has been down for several hours (for example). If the DC goes down and won't come up, the only way is to seize the role, but then the previous owner must never go online again. This seems to me need to be manual interaction, and would be to dangerous to automate. 🤔 

So instead of having a new resource handling FSMO roles I suggest adding this to the ADDomainController by a parameter `FlexibleSingleMasterOperationRole` which is a validate set of the supported roles `'DomainNamingMaster'`, `'SchemaMaster'`, `'InfrastructureMaster'`, `'PDCEmulator'`, `'RIDMaster'`. This way when the domain controller is provisioned and up, on the second run when the configuration is enforced it will move the role(s) that are specified to itself, but only if the current holder is online. This will also result in that if the domain controller get decommissioned then the configuration for that domain controller will also be decommissioned, and a new configuration will be built to set up the new domain controller. In this scenario it would be possible to actually have a parameter `AllowSeizeRole` to forcibly take over a role from a decommissioned domain controller.

On the second run when the node is already a domain controller the resource ADDomainController will do this by first get the current holder of the role(s).

```powershell
PS> Get-ADForest contoso.com | ft DomainNamingMaster, SchemaMaster
DomainNamingMaster  SchemaMaster
------------------  ------------
condc01.contoso.com condc01.contoso.com

PS> Get-ADDomain contoso.com | ft InfrastructureMaster, PDCEmulator, RIDMaster
InfrastructureMaster PDCEmulator         RIDMaster
-------------------- -----------         ---------
condc01.contoso.com  condc01.contoso.com condc01.contoso.com
```

Then use the `Server` parameter to try to connect to current holder (to get an error message it it is down).

```powershell
<# 
    Using Get-ADDomainController to handle a issue with calling
    Move-ADDirectoryServerOperationMasterRole with FQDN
    in the Identity parameter.
#>
$newRoleHolder = Get-ADDomainController -Identity 'newdc01.contoso.com'
Move-ADDirectoryServerOperationMasterRole -Identity $newRoleHolder -OperationMasterRole $FlexibleSingleMasterOperationRole -Server 'condc01.contoso.com'
```

**Comment by jhochwald** [comment]
A logic that takes the best (fastest?) available might be a good idea.

**Comment by matthitchcock** [comment]
I disagree with seizing a FSMO role using this method. Seizing is an exceptional process and should be done with manual intervention. The DSC configuration should then be updated to reflect the change and to apply should the environment be reprovisioned or to cleanly move the role.

Having the DSC configuration fail to move the role if the original role holder is down is the right thing to do. Once manual intervention is done for this important task, the configuration will succeed.

Let’s be clear that DSC is “Desired” state configuration, not “must be able to overcome every possible failure” configuration.

**Comment by johlju** [comment]
@matthitchcock Thank you for your input, I agree with what you are saying. I personally would feel more comfortable if no resource is having the ability to seize.

**Comment by johlju** [comment]
@jhochwald Thanks for your input. Not sure it is practically possible. 🤔 I think we must assume the Active Directory cmdlets/classes have the best logic to give us the best possible domain controller. The domain controller that is "the fastest" might not be the one we want, because it might be fastest for that moment, but not in the next moment then another is "faster" because of various circumstances.

**Comment by jhochwald** [comment]
Hi @johlju I know. That would require a change within the AD concept. This kind of information is something that need a lot of information from the inside.
In such a use case the role model would be fluent and all the master roles would be moved around, based on load and network performance.

**Comment by raandree** [comment]
During a couple of DSC / AD projects the question arose how to manage the FSMO roles with the DC deployment. As changes on the FSMO roles are happening very rarely it did not hurt that there was no converge in the DSC resources yet. But it is definitely nice to have.

I would not want to assign the FSMO roles dynamically considering replication costs or load. It should be a decision that someone who knows the infrastructure does. That decision should be clearly reflected in the config data.

I do agree with @matthitchcock that seizing roles should not be done by DSC. A missing FSMO role is either an unexpected failure on a non-clean demotion. In either way, I would like to have the control of the process finding a new FSMO role owner.

**Comment by johlju** [comment]
@raandree thank you for your input!

**Comment by johlju** [comment]
It sounds like adding a parameter to the resource ADDomainController might be a better approach than having a new resource. WIth a new parameter we get better control where the roles are when they are bound to a certain "DSC role"-configuration (domain controller), than any computer with enough permission could initiate a move if there were a dedicated resource for this. Using resource ADDomainController there need to be a provisioned domain controller before any move can be done, since my suggestion is that it always move to self.

**Comment by johlju** [comment]
Also, seizing should not be allowed. If a move cannot be done, the resource throws an error since the desired state cannot be set.

**Comment by johlju** [comment]
Thank you all for this! I have merged a PR fixing this.
